pkg/common/consumer/consumer_handler.go

```diff
@@ -108,24 +120,50 @@ func (consumer *SaramaConsumerHandler) Cleanup(session sarama.ConsumerGroupSessi
 func (consumer *SaramaConsumerHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
 	consumer.logger.Infow(fmt.Sprintf("Starting partition consumer, topic: %s, partition: %d, initialOffset: %d", claim.Topic(), claim.Partition(), claim.InitialOffset()), zap.String("ConsumeGroup", consumer.handler.GetConsumerGroup()))
 	consumer.handler.SetReady(claim.Partition(), true)
+	c := make(chan bool)
+
 	// NOTE:
 	// Do not move the code below to a goroutine.
 	// The `ConsumeClaim` itself is called within a goroutine, see:
 	// https://github.com/Shopify/sarama/blob/master/consumer_group.go#L27-L29
 	for message := range claim.Messages() {
-
 		if ce := consumer.logger.Desugar().Check(zap.DebugLevel, "debugging"); ce != nil {
 			consumer.logger.Debugw("Message claimed", zap.String("topic", message.Topic), zap.Binary("value", message.Value))
 		}
 
-		// Don't use the session context since it is closed before messages are drained.
-		// Handle must finish before the session timeout.
-		mustMark, err := consumer.handler.Handle(context.Background(), message)
+		// Preemptively interrupt processing messages if the session is closed.
+		// Processing all messages from the buffered channel can take a long time,
+		// potentially exceeding the session timeout, leading to duplicate events.
+		if session.Context().Err() != nil {
+			consumer.logger.Infof("Session closed for %s/%d. Exiting ConsumeClaim ", claim.Topic(), claim.Partition())
+			break
+		}
+
+		hctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			mustMark, err := consumer.handler.Handle(hctx, message)
+
+			if err != nil {
```

Comment:
plesae see here for nested go routine: https://github.com/knative-sandbox/eventing-kafka/pull/659/files#diff-318b63710b5768647bfeb5b9bca7a04b80f3a8388439a6b8bc50eddfb8fed8e7R125-R128

Reply:
That comment is about moving the whole processing in a separate routine, the expectation is that `ConsumeClaim` ideally should block and loop over `Messages` which is the case here, cause we're blocking still in the select on line 156